raft: add MsgDeFortifyLeader

arulajmani opened the pull request:
This patch adds a pair of messages to Raft -- MsgLeaderDeFortify and
MsgDeFortifyLeaderResp. These will be used by a leader, once it has
stepped down, to proactively nudge followers that it can be safely
de-fortified. This then enables followers to campaign themselves or
vote for other peers in an election.

For now, the step down -> de-fortification flow isn't hooked up as a
state transition in raft. We're missing a few things before we can
fully do this:
- The leader needs to ensure it has waited out any LeadSupportUntil
it has communicated to the leasing layer above before it can safely
de-fortify.
- The leader needs to track the leadership term its de-fortifying.

Informs https://github.com/cockroachdb/cockroach/issues/129098

Release note: None

arulajmani committed:
raft: add MsgDeFortifyLeader

This patch adds a new message to Raft, MsgDeFortifyLeader. In a
subsequent patch, this will be used by a once fortified leader to
proactively nudge followers that it can be safely de-fortified. This
then enables followers to campaign themselves or vote for other peers
in an election. The leader will only broadcast this message when it
has stepped down.

Note that it isn't sufficient for the leader to just de-fortify
followers that it was actively tracking fortification state for. That's
because the leader's tracking is pessimistic, in that a follower may be
fortifying a leader unbenknownst to it. To that end, de-fortification
attempts are decoupled from fortification.

Relatedly, when a leader steps down, it also needs to de-fortify
learners. That's because they may have been voters in the previous
configuration, and they may need to vote for a candidate who hasn't
applied the newest configuration yet.

For now, the step down -> defortification flow isn't hooked up as a
state transition in raft. We're missing a few things before we can
fully do this:
- The leader needs to ensure it has waited out any LeadSupportUntil
it has communicated to the leasing layer above before it can safely
de-fortify.
- The leader needs to track the leadership term it's de-fortifying.

Informs https://github.com/cockroachdb/cockroach/issues/129098

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131668)
<!-- Reviewable:end -->

arulajmani committed:
raft: pull logger into its own package

This will allow us to use it in other places, in particular, in the
FortificationTracker.

Epic: none

Release note: None

arulajmani commented:
This should be good for a look now. I've updated the PR after our discussions from earlier today. 

I also pulled out https://github.com/cockroachdb/cockroach/pull/131708 for the first 3 commits on here. Those should be relatively uncontroversial, so landing them separately should help reduce the review burden on this as we iterate on the core logic.

iskettaneh reviewed (commented):
> Reviewed 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 21 of 21 files at r5, 14 of 15 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131668)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`-- commits` line 52 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O89WkFN7Nihc1tUNCSk):*
I am curious what happens if the leader doesn't wait it out? We can discuss it offline
___
*[`pkg/raft/raft.go` line 546 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O89Xa-i1YNGcrYkL4B4:-O89Xa-i1YNGcrYkL4B5:bbz73wa) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L546)):*
> ```Go
> 			//   MsgPreVote if the pre-vote was granted, non-zero for the
> 			//   same reasons MsgPreVote is
> 			//
> ```

Nit: extra line and slightly different format than the messages above
___
*[`pkg/raft/raft.go` line 2316 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O89d8Nf2RL8QUJ9-YfF:-O89d8Nf2RL8QUJ9-YfG:btdaztv) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L2316)):*
> ```Go
> 
> func (r *raft) handleDeFortify(m pb.Message) {
> 	assertTrue(r.state == StateFollower, "leaders should locally de-fortify without sending a message")
> ```

Could also be a candidate right?
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 170 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O89bSg46B8OJ84Korvg:-O89bSg46B8OJ84Korvh:b2wdveu) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/rafttest/interaction_env_handler.go#L170)):*
> ```Go
> 		// step-down 1
> 		err = env.handleStepDown(t, d)
> 	case "maybe-send-de-fortify":
> ```

Is this a temporary command because defortification isn't completely hooked yet?
___
*[`pkg/raft/log_test.go` line 179 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O89VyNRCABh9Wi2uERx:-O89VyNRCABh9Wi2uERy:b-aqh84o) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdc57d844ff49c984493011569d7a86d2adbf290/pkg/raft/log_test.go#L179)):*
> ```Go
> 			require.NoError(t, storage.SetHardState(pb.HardState{Term: init.term}))
> 			require.NoError(t, storage.Append(init.entries))
> 			og := newLog(storage, logger.DiscardLogger)
> ```

what is `og`?
___
*[`pkg/raft/tracker/fortificationtracker_test.go` line 280 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O89cXAI07txZ6dlEQb4:-O89cXAI07txZ6dlEQb5:bwx8dsc) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/tracker/fortificationtracker_test.go#L280)):*
> ```Go
> 				// never learned about this. Maybe we then hear about this when trying
> 				// to de-fortify.
> 				fortificationTracker.RecordDeFortification(1, 11)
> ```

Do we need another test to record defortification for a lower epoch?
___
*[`pkg/raft/logger/logger.go` line 65 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O89VXMb8DJs-zI_QhSX:-O89VXMb8DJs-zI_QhSY:b-kuneg8) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdc57d844ff49c984493011569d7a86d2adbf290/pkg/raft/logger/logger.go#L65)):*
> ```Go
> 	DiscardLogger     = &DefaultLogger{Logger: log.New(io.Discard, "", 0)}
> 	raftLoggerMu      sync.Mutex
> 	RaftLogger        = Logger(DefaultRaftLogger)
> ```

Curious: What's the difference between the DefaultRaftLogger and the RaftLogger?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 21 of 21 files at r5, 7 of 15 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131668)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @iskettaneh)

___
*[`-- commits` line 37 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8ASMnF0Ey-NgVqd1SG):*
`MsgDeFortifyLeader`
___
*[`-- commits` line 41 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8ASUf1AD1wMC5kOYah):*
"peers"
___
*[`-- commits` line 47 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8ASh4yEpJnEN1kPj_e):*
Not needed in this PR, but reminder that we'll also need to defortify learner replicas, because they may have been voter replicas in the previous configuration and they may need to vote for a voter who has not yet applied the newest config. Related to https://github.com/cockroachdb/cockroach/pull/42251.
___
*[`-- commits` line 52 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O89WkFN7Nihc1tUNCSk):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

I am curious what happens if the leader doesn't wait it out? We can discuss it offline
</blockquote></details>

If the leader immediately defortifies, then another leader can immediately step up and then may immediately acquire a lease. This may appear to the old leader/leaseholder as a lease expiration regression. That's not ok, especially because the old leader/leaseholder may have served reads within its lease interval but in the future of its current clock time. We wait out the LeadSupportUntil before defortifying to avoid this.
___
*[`-- commits` line 55 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AU27q770xkzNAUwE1):*
"it's"
___
*[`pkg/raft/raft.go` line 585 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AUmUg107yiOE8ojRt:-O8AUmUg107yiOE8ojRu:bdcv6l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L585)):*
> ```Go
> 		// leader's epoch being supported. The same applies to de-fortifying a
> 		// leader as well, as once the leader has successfully de-fortified a
> 		// follower, it will not try again.
> ```

Good point.
___
*[`pkg/raft/raft.go` line 1431 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AVE0Q0QS5m_zSoBxN:-O8AVE0Q0QS5m_zSoBxO:b-q814t4) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L1431)):*
> ```Go
> 				r.id, r.Term, m.Type, m.From, m.Term)
> 			if IsMsgFromLeader(m.Type) ||
> 				m.Type == pb.MsgDeFortifyLeader {
> ```

Seeing this now, it seems even more like `MsgDeFortifyLeader` should cause `IsMsgFromLeader` to return true. `IsMsgFromLeader` doesn't mean that the message is from the current leader (it can't know that, messages may be delayed), it just means that the message is from the leader of `m.Term`. Why treat it separately (and then need to write this long comment to justify the separate treatment)?
___
*[`pkg/raft/raft.go` line 1497 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8A_sjv3rYM1GWZbKbS:-O8A_sjv3rYM1GWZbKbT:bnhq2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L1497)):*
> ```Go
> 				"%x received %s from [%d, term %d]; de-fortification is a no-op; current term %d",
> 				r.id, pb.MsgDeFortifyLeader, m.From, m.Term, r.Term)
> 			r.send(pb.Message{To: m.From, Type: pb.MsgDeFortifyLeaderResp, LeadEpoch: 0})
> ```

(revisiting based on comment below) this is complexity only needed due to the defortification ack.
___
*[`pkg/raft/raft.go` line 2011 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8Aa22L-3UF_POup058:-O8Aa22L-3UF_POup059:bnhq2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L2011)):*
> ```Go
> 		// wouldn't attempt to de-fortify them in the first place.
> 		assertTrue(r.leadEpoch == 0, "candidates shouldn't be fortified")
> 		r.handleDeFortify(m)
> ```

(revisiting based on comment below) this is complexity only needed due to the defortification ack.
___
*[`pkg/raft/raft.go` line 2320 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AWVPs0jQtMvJefGJU:-O8AWVPs0jQtMvJefGJV:br0gobn) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L2320)):*
> ```Go
> 
> 	leadEpoch := r.leadEpoch
> 	r.deFortify(m.From, m.Term)
> ```

What would happen if a delayed `MsgFortifyLeader` lands on the follower after we have acked the defortification? It seems like the follower would be re-fortified again, but then the ex-leader would never come back to re-de-fortify.

If that's the case, then we have one of two options:
1. don't track successfully defortified followers on the ex-leader and just keep sending defortification messages to them
2. move them to a state where they can't be refortified again

I was thinking that advancing to the next term would be the easiest approach to implement option 2, but we don't want to have all followers immediately campaign. We could probably make this work, but I'm not sure that it's worth the effort.

Instead, I think we should seriously consider option 1 and remove `MsgDeFortifyLeaderResp`. It seems to add a decent amount of complexity to this defortification protocol, and I'm not sure it buys us anything concrete. The argument for it is that we are sending unnecessary messages when an ex-leader could know that it has defortified a follower. That argument is reasonable in theory, but I think it falls apart in practice because:
1. a range that can achieve quorum after a leader stepdown should quickly be able to commit a new entry. A few extra `MsgDeFortifyLeader` broadcasts do almost no harm. They're probably a wash with the cost of a response to each message. Once the next entry is committed, we'll stop defortifying.
2. a range that cannot achieve quorum is already chatty, with replicas periodically broadcasting MsgPreVote and MsgVote. Broadcasting an extra message doesn't meaningfully increase the cost ...
3. especially because if we broadcast at the same time as we campaign, the messages will be batched to each remote peer and we get the `MsgDeFortifyLeader` essentially for free.

I'm sure there are cases where the ability to return a `MsgDeFortifyLeaderResp` helps in some way, but since we don't know how much it helps and since we're contending with real complexity trying to coordinate this handshake in a way that is tolerant of reorderings and tracks the right state, I'd suggest that we consider this a pre-mature optimization for now and put it in our back pocket to add later, if we ever decide that it's needed.

What do you think?
___
*[`pkg/raft/raft.go` line 2338 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AaBqA6nF4iKUAmTJ_:-O8AaBqA6nF4iKUAmTJa:b-buwkf8) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L2338)):*
> ```Go
> func (r *raft) deFortify(from pb.PeerID, term uint64) {
> 	assertTrue(
> 		// We're not currently de-fortified, so de-fortification is a no-op...
> ```

"not currently fortified
___
*[`pkg/raft/logger/BUILD.bazel` line 4 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AQy48BeKgd65eTbbB:-O8AQy48BeKgd65eTbbC:b-367dr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdc57d844ff49c984493011569d7a86d2adbf290/pkg/raft/logger/BUILD.bazel#L4)):*
> ```Starlark
> 
> go_library(
>     name = "logger",
> ```

I have a slight preference for `pkg/raft/raftlog` and `raftlog.Logger` so that the relationship between this package and `pkg/util/log` is more clear when used in code. What do you think?
___
*[`pkg/raft/tracker/fortificationtracker.go` line 102 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AWBsVBkflr2QXTLQP:-O8AWBsVBkflr2QXTLQQ:b-7vr0km) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/tracker/fortificationtracker.go#L102)):*
> ```Go
> 	// follower independently de-fortified because it was safe to do so. Make sure
> 	// something breaks before adding that conditional.
> 	if epoch < curEpoch {
> ```

Was the point of shipping the epoch back to the ex-leader just so that we can perform this assertion? Does the assertion provide any protection?
___
*[`pkg/raft/log_test.go` line 179 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O89VyNRCABh9Wi2uERx:-O8ARcfTBAxzQGcUtG0t:b4dmd1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdc57d844ff49c984493011569d7a86d2adbf290/pkg/raft/log_test.go#L179)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

what is `og`?
</blockquote></details>

Good catch. I don't think this was intentional.
___
*[`pkg/raft/tracker/fortificationtracker_test.go` line 198 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AQkEgAjJuTUoIZHwT:-O8AQkEgAjJuTUoIZHwU:b-3jrmnl) ([raw file](https://github.com/cockroachdb/cockroach/blob/86a7423493906bcb06e97075cf87356248e02c6b/pkg/raft/tracker/fortificationtracker_test.go#L198)):*
> ```Go
> 			// No support recorded at the store liveness fabric.
> 			storeLiveness: makeMockStoreLiveness(map[pb.PeerID]mockLivenessEntry{}),
> 			setup: func(fortificationTracker *FortificationTracker) {
> ```

nit: should we just use `ft *FortificationTracker` throughout this test file?
___
*[`pkg/raft/logger/logger.go` line 65 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O89VXMb8DJs-zI_QhSX:-O8ARIwn-uDIsHGSfP1G:b809ow8) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdc57d844ff49c984493011569d7a86d2adbf290/pkg/raft/logger/logger.go#L65)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Curious: What's the difference between the DefaultRaftLogger and the RaftLogger?
</blockquote></details>

The difference is that `RaftLogger` can be overridden by `SetLogger`. `defaultLogger` is meant to be const.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
RFAL. Now that we don't have `MsgDeFortifyLeaderResp`, the first commit (the one that pulls the logger into its own package), is no longer necessary. That's because I'm not logging anything in the fortification tracker. I think it's still worth keeping -- I've previously wished I could use the logger in the `tracker` package, and the commit enables us to do that in the future.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-:-O8HmGp3EqaQWPaGUPKG:bmt8uj6)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh and @nvanbenschoten)

___
*[`-- commits` line 47 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8ASh4yEpJnEN1kPj_e):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Not needed in this PR, but reminder that we'll also need to defortify learner replicas, because they may have been voter replicas in the previous configuration and they may need to vote for a voter who has not yet applied the newest config. Related to https://github.com/cockroachdb/cockroach/pull/42251.
</blockquote></details>

Added some words.
___
*[`pkg/raft/raft.go` line 546 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O89Xa-i1YNGcrYkL4B4:-O8HeiSA6hLBVgrA3klX:b41jnha) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L546)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Nit: extra line and slightly different format than the messages above
</blockquote></details>

Got caught in a rebase. Thanks for catching, fixed.
___
*[`pkg/raft/raft.go` line 585 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AUmUg107yiOE8ojRt:-O8Hf7KQB7yt-FBxXdZt:bikfmen) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L585)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Good point.
</blockquote></details>

Not after we get rid of `MsgDeFortifyResp` :joy: 
___
*[`pkg/raft/raft.go` line 1431 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AVE0Q0QS5m_zSoBxN:-O8Hff0-22K8Yhb8bw1p:b-lseb9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L1431)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Seeing this now, it seems even more like `MsgDeFortifyLeader` should cause `IsMsgFromLeader` to return true. `IsMsgFromLeader` doesn't mean that the message is from the current leader (it can't know that, messages may be delayed), it just means that the message is from the leader of `m.Term`. Why treat it separately (and then need to write this long comment to justify the separate treatment)?
</blockquote></details>

I think what tripped me was the "we've just received a message from the new leader". I'm going to switch the "the" to a "a". 

I've reworded a bit of the comment instead of removing it entirely. I can remove it too, if it feels like information overload for no good reason.
___
*[`pkg/raft/raft.go` line 1497 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8A_sjv3rYM1GWZbKbS:-O8HhGOLAGYv_ca56_K5:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L1497)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

(revisiting based on comment below) this is complexity only needed due to the defortification ack.
</blockquote></details>

Removed.
___
*[`pkg/raft/raft.go` line 2011 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8Aa22L-3UF_POup058:-O8HhHOHBEWo1DrSUgj5:b-v6s1zr) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L2011)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

(revisiting based on comment below) this is complexity only needed due to the defortification ack.
</blockquote></details>

Removed .
___
*[`pkg/raft/raft.go` line 2316 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O89d8Nf2RL8QUJ9-YfF:-O8Df0iD5vJaHVbt0n6V:b-xc22wn) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L2316)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Could also be a candidate right?
</blockquote></details>

Yeah, if it's in its `PreVote` phase (otherwise, it would've bumped its term and we'd never get here). Fixed.
___
*[`pkg/raft/raft.go` line 2320 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AWVPs0jQtMvJefGJU:-O8Hje5j2hLBHXmUDsWr:bqeg1fs) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L2320)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What would happen if a delayed `MsgFortifyLeader` lands on the follower after we have acked the defortification? It seems like the follower would be re-fortified again, but then the ex-leader would never come back to re-de-fortify.

If that's the case, then we have one of two options:
1. don't track successfully defortified followers on the ex-leader and just keep sending defortification messages to them
2. move them to a state where they can't be refortified again

I was thinking that advancing to the next term would be the easiest approach to implement option 2, but we don't want to have all followers immediately campaign. We could probably make this work, but I'm not sure that it's worth the effort.

Instead, I think we should seriously consider option 1 and remove `MsgDeFortifyLeaderResp`. It seems to add a decent amount of complexity to this defortification protocol, and I'm not sure it buys us anything concrete. The argument for it is that we are sending unnecessary messages when an ex-leader could know that it has defortified a follower. That argument is reasonable in theory, but I think it falls apart in practice because:
1. a range that can achieve quorum after a leader stepdown should quickly be able to commit a new entry. A few extra `MsgDeFortifyLeader` broadcasts do almost no harm. They're probably a wash with the cost of a response to each message. Once the next entry is committed, we'll stop defortifying.
2. a range that cannot achieve quorum is already chatty, with replicas periodically broadcasting MsgPreVote and MsgVote. Broadcasting an extra message doesn't meaningfully increase the cost ...
3. especially because if we broadcast at the same time as we campaign, the messages will be batched to each remote peer and we get the `MsgDeFortifyLeader` essentially for free.

I'm sure there are cases where the ability to return a `MsgDeFortifyLeaderResp` helps in some way, but since we don't know how much it helps and since we're contending with real complexity trying to coordinate this handshake in a way that is tolerant of reorderings and tracks the right state, I'd suggest that we consider this a pre-mature optimization for now and put it in our back pocket to add later, if we ever decide that it's needed.

What do you think?
</blockquote></details>

We discussed this offline. Entirely onboard with this -- cuts down the complexity massively, as we no longer need to do any de-fortification tracking on the ex-leader. 
___
*[`pkg/raft/raft.go` line 2338 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AaBqA6nF4iKUAmTJ_:-O8HhMww4LdYjjHcH_lt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L2338)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"not currently fortified
</blockquote></details>

Done.
___
*[`pkg/raft/logger/BUILD.bazel` line 4 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AQy48BeKgd65eTbbB:-O8DbsDp5k16n3VaTP8U:b-o4fd5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdc57d844ff49c984493011569d7a86d2adbf290/pkg/raft/logger/BUILD.bazel#L4)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I have a slight preference for `pkg/raft/raftlog` and `raftlog.Logger` so that the relationship between this package and `pkg/util/log` is more clear when used in code. What do you think?
</blockquote></details>

`raftlog` is a bit ambiguous, in that it's not clear whether it means the raft log or a logging utility inside the raft package. Maybe it's just me, but if I see "log" inside this package, I assume it's the former. Adding the "er" to make it "logger" helps disambiguate these for me. 

Would you prefer if we called this package `raftlogger` instead? 
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 170 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O89bSg46B8OJ84Korvg:-O8HhPBv9nVnBa0rmiCM:bcaa37x) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/rafttest/interaction_env_handler.go#L170)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Is this a temporary command because defortification isn't completely hooked yet?
</blockquote></details>

I was planning on keeping it around even then, but I could be convinced otherwise. Mostly to test the cases where de-fortification is a no-op because the follower has already been de-fortified. That being said, I think we can achieve that by just ticking an election, so maybe we can port over `de_fortification_basic.txt` and get rid of this entirely once things are hooked up fully.
___
*[`pkg/raft/tracker/fortificationtracker.go` line 102 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AWBsVBkflr2QXTLQP:-O8HirBzCl34JZl7gV5w:b-e1bv0w) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/tracker/fortificationtracker.go#L102)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Was the point of shipping the epoch back to the ex-leader just so that we can perform this assertion? Does the assertion provide any protection?
</blockquote></details>

Yeah, back when I hadn't introduced this `deFortification` map and I was just removing things from `ft.fortification`, I had this sanity check. Wasn't critical in any sense, and doesn't apply now that we don't have any de-fortification tracking/`MsgDeFortifyLeader`.
___
*[`pkg/raft/log_test.go` line 179 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O89VyNRCABh9Wi2uERx:-O8DbX_LEpkCDwao75AI:bo3e82a) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdc57d844ff49c984493011569d7a86d2adbf290/pkg/raft/log_test.go#L179)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Good catch. I don't think this was intentional.
</blockquote></details>

Not entirely sure how that happened. Thanks for catching; reverted.
___
*[`pkg/raft/tracker/fortificationtracker_test.go` line 198 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AQkEgAjJuTUoIZHwT:-O8HjQ-aAz92NeyLZFD9:b28ncs0) ([raw file](https://github.com/cockroachdb/cockroach/blob/86a7423493906bcb06e97075cf87356248e02c6b/pkg/raft/tracker/fortificationtracker_test.go#L198)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: should we just use `ft *FortificationTracker` throughout this test file?
</blockquote></details>

Merged this in a different PR, so I'll defer acting on this for now.
___
*[`pkg/raft/tracker/fortificationtracker_test.go` line 280 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O89cXAI07txZ6dlEQb4:-O8HjWRj9MfIt_SuqHzb:b-w470vl) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/tracker/fortificationtracker_test.go#L280)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Do we need another test to record defortification for a lower epoch?
</blockquote></details>

No longer applies now that we don't have a concept of recording de-fortifications.


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (approved):
> Reviewed 18 of 18 files at r7, 5 of 15 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131668)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 367 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8IM6JW0oMw7G8QTgiO:-O8IM6JW0oMw7G8QTgiP:b65pey) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4b893ca97fff8b8956292fcc2d1bbb04bc57d35/pkg/raft/raft.go#L367)):*
> ```Go
> 	//
> 	// Messages in this list have the type MsgAppResp, MsgVoteResp,
> 	// MsgPreVoteResp, MsgFortifyLeaderResp, or MsgDeFortifyLeaderResp. See the
> ```

We removed MsgDeFortifyLeaderResp from this PR right?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 1 of 18 files at r7, 7 of 15 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131668)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @iskettaneh)

___
*[`pkg/raft/raft.go` line 1431 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AVE0Q0QS5m_zSoBxN:-O8IppST26oPML0GuL0Q:bsvpsce) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L1431)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I think what tripped me was the "we've just received a message from the new leader". I'm going to switch the "the" to a "a". 

I've reworded a bit of the comment instead of removing it entirely. I can remove it too, if it feels like information overload for no good reason.
</blockquote></details>

We could even say "a message from a leader". It doesn't have to be new.

Do we want to push MsgDeFortifyLeader into `IsMsgFromLeader`? EDIT: actually, you already did. So we can just get rid of the `|| m.Type == pb.MsgDeFortifyLeader`
___
*[`pkg/raft/raft.go` line 2011 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8Aa22L-3UF_POup058:-O8Iq8b3AVRwj6bQQNPU:b-x1qq8c) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L2011)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Removed .
</blockquote></details>

Check again :smiley: 
___
*[`pkg/raft/logger/BUILD.bazel` line 4 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AQy48BeKgd65eTbbB:-O8IpI0Y-IpsLiTSOBBk:b-qttvg8) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdc57d844ff49c984493011569d7a86d2adbf290/pkg/raft/logger/BUILD.bazel#L4)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

`raftlog` is a bit ambiguous, in that it's not clear whether it means the raft log or a logging utility inside the raft package. Maybe it's just me, but if I see "log" inside this package, I assume it's the former. Adding the "er" to make it "logger" helps disambiguate these for me. 

Would you prefer if we called this package `raftlogger` instead?
</blockquote></details>

Great point, I had missed how bad of a name that would be. `raftlogger` SGTM.
___
*[`pkg/raft/tracker/fortificationtracker_test.go` line 198 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AQkEgAjJuTUoIZHwT:-O8IpVEL5D-NPWAiaYL2:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/86a7423493906bcb06e97075cf87356248e02c6b/pkg/raft/tracker/fortificationtracker_test.go#L198)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Merged this in a different PR, so I'll defer acting on this for now.
</blockquote></details>

:thumbsup: 


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-:-O8JJNVA21KmPL6vbCoJ:bb72pd0)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 1431 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AVE0Q0QS5m_zSoBxN:-O8JEeGgDAkr7qkfAs28:b8yydxt) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L1431)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We could even say "a message from a leader". It doesn't have to be new.

Do we want to push MsgDeFortifyLeader into `IsMsgFromLeader`? EDIT: actually, you already did. So we can just get rid of the `|| m.Type == pb.MsgDeFortifyLeader`
</blockquote></details>

Ah sorry, had meant to remove it. Fixed.
___
*[`pkg/raft/raft.go` line 2011 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8Aa22L-3UF_POup058:-O8JGGZh-kKBEzzX001_:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L2011)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Check again :smiley:
</blockquote></details>

Fixed.
___
*[`pkg/raft/raft.go` line 367 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8IM6JW0oMw7G8QTgiO:-O8JETaOAieN0HcQBiOO:b-r5d6ua) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4b893ca97fff8b8956292fcc2d1bbb04bc57d35/pkg/raft/raft.go#L367)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

We removed MsgDeFortifyLeaderResp from this PR right?
</blockquote></details>

We did. Fixed.
___
*[`pkg/raft/logger/BUILD.bazel` line 4 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8AQy48BeKgd65eTbbB:-O8JJJMG4eD8UAMQ5r1V:b-6au0n6) ([raw file](https://github.com/cockroachdb/cockroach/blob/fdc57d844ff49c984493011569d7a86d2adbf290/pkg/raft/logger/BUILD.bazel#L4)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Great point, I had missed how bad of a name that would be. `raftlogger` SGTM.
</blockquote></details>

Switched to `raftlogger`.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> once the remaining threads are addressed.

> Reviewed 1 of 15 files at r8, 13 of 32 files at r9, 12 of 15 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-:-O8Kglyv4SyTqZRNjCha:b-gs78u5)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @iskettaneh)

___
*[`pkg/raft/raft.go` line 2011 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8Aa22L-3UF_POup058:-O8Kd4AV0lU9dFZwYqPP:b58bbvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L2011)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Fixed.
</blockquote></details>

I really feel like we should just delete this. We shouldn't need a message handler just to perform a local assertion that's unrelated to the incoming message. If we want to assert that `r.leadEpoch == 0` while in `StateCandidate` or `StatePreCandidate`, we should make that assertion during the `r.state` transition, not here.

These kinds of "might as well" assertions aren't terribly costly in terms of complexity when written, but they can become more costly over time, as code gets restructured. For example, maybe we want to add a catch-all "message dropped" branch with some custom logging at some later point. Then `MsgDeFortifyLeader` gets missed. Or worse, we then need to start duplicating code.
___
*[`pkg/raft/raft.go` line 367 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8IM6JW0oMw7G8QTgiO:-O8KeYgKAnTO3dvQ2Gx_:b-i42h1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4b893ca97fff8b8956292fcc2d1bbb04bc57d35/pkg/raft/raft.go#L367)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

We did. Fixed.
</blockquote></details>

We're now missing the "or". We just want to revert the entire diff in this comment.
___
*[`pkg/raft/raft.go` line 1430 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8KexqE-OcDi490hmKM:-O8KexqE-OcDi490hmKN:b-mkox0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/13eda013ba5e4e58e01835abe6a32897bdcf23a2/pkg/raft/raft.go#L1430)):*
> ```Go
> 				// from a MsgDeFortifyLeader.
> 				//
> 				// [1] A leader must broadcast MsgDeFortifyLeader to all peers when it
> ```

I don't feel strongly about this so feel free to ignore, but this third paragraph feels out of place. Even the second paragraph is arguably redundant with the first, but this third one in particular feels like an unimportant implementation detail that this code doesn't need to worry about. Including it creates noise and requires each reader to question whether it's necessary to understand in order to understand the code. In some ways, it feels like justification for some of the complexity of the stateful approach which we've now moved away from.

The fact that a leader must broadcast MsgDeFortifyLeader to all peers when it steps down should definitely be documented near the logic that performs the MsgDeFortifyLeader broadcast, just not here.
___
*[`pkg/raft/util.go` line 60 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8KgZ-F8KucViMemvyc:-O8KgZ-F8KucViMemvyd:b-w8gk8r) ([raw file](https://github.com/cockroachdb/cockroach/blob/13eda013ba5e4e58e01835abe6a32897bdcf23a2/pkg/raft/util.go#L60)):*
> ```Go
> 	pb.MsgSnap:            true,
> 	pb.MsgHeartbeat:       true,
> 	pb.MsgFortifyLeader:   true,
> ```

small nit: it's on me, but while you're here, mind ordering these correctly so that `MsgFortifyLeader` and `MsgDeFortifyLeader` are next to each other?
___
*[`pkg/raft/raftpb/raft.proto` line 67 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8KfogsAQHiJTZbKak2:-O8KfogsAQHiJTZbKak3:b-n4uo8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/13eda013ba5e4e58e01835abe6a32897bdcf23a2/pkg/raft/raftpb/raft.proto#L67)):*
> ```proto
>   MsgFortifyLeader       = 24;
>   MsgFortifyLeaderResp   = 25;
>   MsgDeFortifyLeader     = 26;
> ```

nit that jumps out from the github diff: now that we removed `MsgDeFortifyLeaderResp`, we don't need the additional indentation and to incur a diff on every variant in the enum. Save yourself from being hit with the `git blame`!


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
TFTR!

Addressed all the comments and fixed the linter. 

bors r=nvanbenschoten,iskettaneh

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-:-O8NckC49PRs4o020DY6:behh2la)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @iskettaneh and @nvanbenschoten)

___
*[`pkg/raft/raft.go` line 2011 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8Aa22L-3UF_POup058:-O8N4EIJ-3EkdHpYDnSB:b44oqgy) ([raw file](https://github.com/cockroachdb/cockroach/blob/82f9e8f1a5fbf9a1c7a91aa4a2dd0a3a0488beba/pkg/raft/raft.go#L2011)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I really feel like we should just delete this. We shouldn't need a message handler just to perform a local assertion that's unrelated to the incoming message. If we want to assert that `r.leadEpoch == 0` while in `StateCandidate` or `StatePreCandidate`, we should make that assertion during the `r.state` transition, not here.

These kinds of "might as well" assertions aren't terribly costly in terms of complexity when written, but they can become more costly over time, as code gets restructured. For example, maybe we want to add a catch-all "message dropped" branch with some custom logging at some later point. Then `MsgDeFortifyLeader` gets missed. Or worse, we then need to start duplicating code.
</blockquote></details>

I misunderstood what you were asking for in your original comment. Fair points, I've removed this case.
___
*[`pkg/raft/raft.go` line 367 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8IM6JW0oMw7G8QTgiO:-O8N3SocDl2I_2psXYHN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4b893ca97fff8b8956292fcc2d1bbb04bc57d35/pkg/raft/raft.go#L367)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We're now missing the "or". We just want to revert the entire diff in this comment.
</blockquote></details>

Done.
___
*[`pkg/raft/raft.go` line 1430 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8KexqE-OcDi490hmKM:-O8N3ad4-Vf2ZoScRPuR:bkr3gam) ([raw file](https://github.com/cockroachdb/cockroach/blob/13eda013ba5e4e58e01835abe6a32897bdcf23a2/pkg/raft/raft.go#L1430)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I don't feel strongly about this so feel free to ignore, but this third paragraph feels out of place. Even the second paragraph is arguably redundant with the first, but this third one in particular feels like an unimportant implementation detail that this code doesn't need to worry about. Including it creates noise and requires each reader to question whether it's necessary to understand in order to understand the code. In some ways, it feels like justification for some of the complexity of the stateful approach which we've now moved away from.

The fact that a leader must broadcast MsgDeFortifyLeader to all peers when it steps down should definitely be documented near the logic that performs the MsgDeFortifyLeader broadcast, just not here.
</blockquote></details>

I'll remove both these paragraphs. I wasn't fully sold on keeping this when I was rewording the comment yesterday as well.
___
*[`pkg/raft/util.go` line 60 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8KgZ-F8KucViMemvyc:-O8N4hGk-Dkh-MwJzhNe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13eda013ba5e4e58e01835abe6a32897bdcf23a2/pkg/raft/util.go#L60)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

small nit: it's on me, but while you're here, mind ordering these correctly so that `MsgFortifyLeader` and `MsgDeFortifyLeader` are next to each other?
</blockquote></details>

Done.
___
*[`pkg/raft/raftpb/raft.proto` line 67 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/131668#-O8KfogsAQHiJTZbKak2:-O8N4zJ_1KBzORhrleoF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13eda013ba5e4e58e01835abe6a32897bdcf23a2/pkg/raft/raftpb/raft.proto#L67)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit that jumps out from the github diff: now that we removed `MsgDeFortifyLeaderResp`, we don't need the additional indentation and to incur a diff on every variant in the enum. Save yourself from being hit with the `git blame`!
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
